File: framework/languages/node/samples/TicTacToe.Ts/Client/tictactoe-client-scenario.ts
import {
  GameMarks,
  GameStatus,
  LeaveGameReq,
  ObserveMilestoneReq,
  PacketNames,
  authenticateReq,
  createGameHttpReq,
  joinGameReq,
  placeMarkStreamReq
} from '../Shared/Contracts/messages';
import { BrowserHttpClientFactory } from '../../browser-client-runtime';
import * as connector from '@zlink-systems/stream-connector';
import type {
  AuthenticateRes,
  CreateGameHttpRes,
  GameState,
  GameStateNotify,
  JoinGameRes,
  ObserveMilestoneRes,
  PlaceMarkRes,
  PlayerJoinedNotify,
  WinMilestoneNotify
} from '../Shared/Contracts/messages';
import type { ZlinkStreamConnector, ZlinkStreamMessage } from '@zlink-systems/stream-connector';

class TicTacToeClientScenario {
  async run(apiHttpEndpoint: string, signal?: AbortSignal): Promise<void> {
    const api = BrowserHttpClientFactory.create(apiHttpEndpoint).build();
    let game: CreateGameHttpRes;
    try {
      // 1. Create the room through API.
      game = await api
        .post('/games')
        .body(createGameHttpReq('match-ready'))
        .fetch<CreateGameHttpRes>();
    } finally {
      await api.close();
    }

    ensure(() => game.gameName === 'match-ready');
    ensure(() => game.roomId.length > 0);
    ensure(() => game.ownerPlayEndpoint.length > 0);
    ensure(() => game.playEndpoints.length >= 2);
    ensure(() => new Set(game.playEndpoints).size === game.playEndpoints.length);
    ensure(() => game.playEndpoints.includes(game.ownerPlayEndpoint));
    ensure(() => game.playNodes.length === game.playEndpoints.length);
    ensure(() => new Set(game.playNodes.map((node) => node.spotNodeRid)).size === game.playNodes.length);
    ensure(() => game.playNodes.every((node) => game.playEndpoints.includes(node.streamEndpoint)));
    ensure(() => game.requiredLevel === 3);
    const ownerPlayNode = game.playNodes.find((node) => node.streamEndpoint === game.ownerPlayEndpoint);
    ensure(() => ownerPlayNode !== undefined);
    ensure(() => ownerPlayNode?.spotNodeRid.length !== 0);

    const guestPlayEndpoint = game.playEndpoints.find((endpoint) => endpoint !== game.ownerPlayEndpoint);
    ensure(() => guestPlayEndpoint !== undefined);
    const observerPlayEndpoint = guestPlayEndpoint as string;
    const observerPlayNode = game.playNodes.find((node) => node.streamEndpoint === observerPlayEndpoint);
    ensure(() => observerPlayNode !== undefined);

    const client1 = createPlayerClient(game.ownerPlayEndpoint, 'host');
    const client2 = createPlayerClient(observerPlayEndpoint, 'guest');
    const observer = createPlayerClient(observerPlayEndpoint, 'observer');

    try {
      // 2. Host, guest, and observer connect directly to Play stream endpoints from the API response.
      await client1.connect(signal);
      const client1Auth = await client1.request(authenticateReq('player-x')).submit<AuthenticateRes>(signal);
      ensure(() => client1Auth.player.actorId === 'player-x');
      ensure(() => client1Auth.player.displayName.length > 0);
      ensure(() => client1Auth.player.level >= game.requiredLevel);
      ensure(() => client1Auth.player.wins === 99);

      await client2.connect(signal);
      const client2Auth = await client2.request(authenticateReq('player-o')).submit<AuthenticateRes>(signal);
      ensure(() => client2Auth.player.actorId === 'player-o');
      ensure(() => client2Auth.player.actorId !== client1Auth.player.actorId);
      ensure(() => client2Auth.player.displayName.length > 0);
      ensure(() => client2Auth.player.level >= game.requiredLevel);
      ensure(() => client2Auth.player.wins >= 0);

      await observer.connect(signal);
      const observerAuth = await observer.request(authenticateReq('observer')).submit<AuthenticateRes>(signal);
      ensure(() => observerAuth.player.actorId === 'observer');
      ensure(() => observerAuth.player.displayName.length > 0);
      ensure(() => observerAuth.player.level >= game.requiredLevel);
      ensure(() => observerAuth.player.wins === 0);
      const observerSubscription = await observer
        .request(new ObserveMilestoneReq())
        .submit<ObserveMilestoneRes>(signal);
      ensure(() => observerSubscription.subscribed);
      console.log('observer-subscription=verified');

      // 3. Host joins by explicit RoomId
      const client1Join = await client1.request(joinGameReq(game.roomId)).submit<JoinGameRes>(signal);
      ensure(() => stateOf(client1Join).roomId === game.roomId);
      ensure(() => stateOf(client1Join).status === GameStatus.WaitingForPlayers);
      ensure(() => stateOf(client1Join).xActorId === client1Auth.player.actorId);
      ensure(() => stateOf(client1Join).oActorId === null);
      ensure(() => stateOf(client1Join).board === '.........');
      await expectNoMessage(
        client1,
        PacketNames.playerJoinedNotify,
        (message: ZlinkStreamMessage<PlayerJoinedNotify>) => message.payload.actorId === client1Auth.player.actorId,
        signal
      );

      const client1SawClient2Join = client1
        .waitFor<PlayerJoinedNotify>(PacketNames.playerJoinedNotify)
        .where((message) => message.payload.actorId === client2Auth.player.actorId)
        .submit(signal);
      const client1RunningState = client1
        .waitFor<GameStateNotify>(PacketNames.gameStateNotify)
        .where((message) => message.payload.state.status === GameStatus.InProgress)
        .submit(signal);

      // 4-6. Guest joins by the same RoomId.
      const client2Join = await client2.request(joinGameReq(game.roomId)).submit<JoinGameRes>(signal);
      ensure(() => stateOf(client2Join).roomId === game.roomId);
      ensure(() => stateOf(client2Join).status === GameStatus.InProgress);
      ensure(() => stateOf(client2Join).oActorId === client2Auth.player.actorId);
      ensure(() => stateOf(client2Join).xActorId === client1Auth.player.actorId);
      ensure(() => stateOf(client2Join).nextTurn === GameMarks.x);
      await expectNoMessage(
        client2,
        PacketNames.playerJoinedNotify,
        (message: ZlinkStreamMessage<PlayerJoinedNotify>) => message.payload.actorId === client2Auth.player.actorId,
        signal
      );

      const [client1Joined, client1Running] = await Promise.all([
        client1SawClient2Join,
        client1RunningState
      ]);
      ensure(() => client1Joined.payload.roomId === game.roomId);
      ensure(() => client1Joined.payload.actorId === client2Auth.player.actorId);
      ensure(() => client1Joined.payload.displayName === client2Auth.player.displayName);
      ensure(() => client1Joined.payload.level === client2Auth.player.level);
      ensure(() => client1Joined.payload.mark === GameMarks.o);
      ensure(() => stateOf(client1Joined.payload).status === GameStatus.InProgress);
      ensure(() => client1Running.payload.state.nextTurn === GameMarks.x);

      // 7. Each move response is matched with the opponent notify.
      const client2SawMove1 = waitState(client2, 0, signal);
      const client1Move1 = await client1.request(placeMarkStreamReq(0)).submit<PlaceMarkRes>(signal);
      requireSameState(stateOf(client1Move1), (await client2SawMove1).payload.state);
      ensure(() => stateOf(client1Move1).board === 'X........');
      ensure(() => stateOf(client1Move1).nextTurn === GameMarks.o);
      requireLastMove(stateOf(client1Move1), client1Auth.player.actorId, 0);

      const client1SawMove2 = waitState(client1, 3, signal);
      const client2Move1 = await client2.request(placeMarkStreamReq(3)).submit<PlaceMarkRes>(signal);
      requireSameState(stateOf(client2Move1), (await client1SawMove2).payload.state);
      ensure(() => stateOf(client2Move1).board === 'X..O.....');
      ensure(() => stateOf(client2Move1).nextTurn === GameMarks.x);
      requireLastMove(stateOf(client2Move1), client2Auth.player.actorId, 3);

      const client2SawMove3 = waitState(client2, 1, signal);
      const client1Move2 = await client1.request(placeMarkStreamReq(1)).submit<PlaceMarkRes>(signal);
      requireSameState(stateOf(client1Move2), (await client2SawMove3).payload.state);
      ensure(() => stateOf(client1Move2).board === 'XX.O.....');
      ensure(() => stateOf(client1Move2).nextTurn === GameMarks.o);
      requireLastMove(stateOf(client1Move2), client1Auth.player.actorId, 1);

      const client1SawMove4 = waitState(client1, 4, signal);
      const client2Move2 = await client2.request(placeMarkStreamReq(4)).submit<PlaceMarkRes>(signal);
      requireSameState(stateOf(client2Move2), (await client1SawMove4).payload.state);
      ensure(() => stateOf(client2Move2).board === 'XX.OO....');
      ensure(() => stateOf(client2Move2).nextTurn === GameMarks.x);
      requireLastMove(stateOf(client2Move2), client2Auth.player.actorId, 4);

      const client2SawFinalMove = client2
        .waitFor<GameStateNotify>(PacketNames.gameStateNotify)
        .where((message) =>
          message.payload.state.roomId === game.roomId &&
          message.payload.state.status === GameStatus.Won &&
          message.payload.state.winner === client1Auth.player.actorId)
        .submit(signal);
      const observerSawMilestone = observer
        .waitFor<WinMilestoneNotify>(PacketNames.winMilestoneNotify)
        .where((message) =>
          message.payload.roomId === game.roomId &&
          message.payload.actorId === client1Auth.player.actorId)
        .submit(signal);
      // 8. The final host move wins and publishes the observer milestone.
      const client1FinalMove = await client1.request(placeMarkStreamReq(2)).submit<PlaceMarkRes>(signal);
      requireSameState(stateOf(client1FinalMove), (await client2SawFinalMove).payload.state);
      ensure(() => stateOf(client1FinalMove).board === 'XXXOO....');
      ensure(() => stateOf(client1FinalMove).status === GameStatus.Won);
      ensure(() => stateOf(client1FinalMove).winner === client1Auth.player.actorId);
      ensure(() => stateOf(client1FinalMove).nextTurn === '');
      requireLastMove(stateOf(client1FinalMove), client1Auth.player.actorId, 2);

      const milestone = await observerSawMilestone;
      ensure(() => milestone.payload.roomId === game.roomId);
      ensure(() => milestone.payload.actorId === client1Auth.player.actorId);
      ensure(() => milestone.payload.displayName === client1Auth.player.displayName);
      ensure(() => milestone.payload.wins === 100);
      ensure(() => milestone.payload.receivingSpotNodeRid === observerPlayNode?.spotNodeRid);
      console.log(
        `observer-win-milestone=verified actor=${milestone.payload.actorId} ` +
        `wins=${milestone.payload.wins} receivingSpotNodeRid=${milestone.payload.receivingSpotNodeRid}`
      );

      await Promise.all([
        client1.send(new LeaveGameReq(game.roomId)).packetName(PacketNames.leaveGameReq).submit(),
        client2.send(new LeaveGameReq(game.roomId)).packetName(PacketNames.leaveGameReq).submit()
      ]);
    } finally {
      await Promise.allSettled([client1.close(), client2.close(), observer.close()]);
    }
  }
}

function waitState(
  client: ZlinkStreamConnector,
  lastMoveCell: number,
  signal?: AbortSignal
): Promise<ZlinkStreamMessage<GameStateNotify>> {
  return client
    .waitFor<GameStateNotify>(PacketNames.gameStateNotify)
    .where((message) => message.payload.state.lastMoveCell === lastMoveCell)
    .submit(signal);
}

async function expectNoMessage<TPayload>(
  client: ZlinkStreamConnector,
  packetName: string,
  predicate: (message: ZlinkStreamMessage<TPayload>) => boolean,
  signal?: AbortSignal
): Promise<void> {
  try {
    await client
      .waitFor<TPayload>(packetName)
      .where(predicate)
      .timeout(25)
      .submit(signal);
  } catch (error) {
    if (isRequestTimeout(error)) {
      return;
    }
    throw error;
  }
  throw new Error(`Unexpected stream message '${packetName}'.`);
}

function isRequestTimeout(error: unknown): boolean {
  return typeof error === 'object'
    && error !== null
    && 'error' in error
    && (error as { error?: { code?: unknown } }).error?.code === 'requestTimeout';
}

function createPlayerClient(endpoint: string, name: string): ZlinkStreamConnector {
  const client = connector.zlinkStreamConnectorFactory.create({
    endpoint,
    dispatchMode: connector.ZlinkStreamDispatchMode.Immediate,
    waitTimeoutMs: 60000,
    heartbeat: { enabled: false }
  });
  client.observeInbound((observation) => {
    console.log(
      `stream-inbound sample=TicTacToe client=${name} kind=${observation.kind} ` +
      `name=${observation.name} seq=${observation.requestSeq?.toString() ?? '-'} ` +
      `bytes=${observation.payloadLength}`
    );
  });
  return client;
}

function stateOf(message: { state: GameState }): GameState {
  return message.state;
}

function requireLastMove(state: GameState, actorId: string, cell: number): void {
  ensure(() => state.lastMoveActorId === actorId);
  ensure(() => state.lastMoveCell === cell);
}

function requireSameState(actual: GameState, expected: GameState): void {
  ensure(() => actual.roomId === expected.roomId);
  ensure(() => actual.board === expected.board);
  ensure(() => actual.status === expected.status);
  ensure(() => actual.winner === expected.winner);
  ensure(() => actual.nextTurn === expected.nextTurn);
  ensure(() => actual.lastMoveActorId === expected.lastMoveActorId);
  ensure(() => actual.lastMoveCell === expected.lastMoveCell);
}

function ensure(condition: () => boolean): void {
  if (!condition()) {
    throw new Error(`Ensure failed: ${conditionExpression(condition)}`);
  }
}

function conditionExpression(condition: () => boolean): string {
  return condition
    .toString()
    .replace(/^\s*\(\)\s*=>\s*/, '')
    .replace(/^\s*function\s*\(\)\s*\{\s*return\s*/, '')
    .replace(/;?\s*\}\s*$/, '')
    .trim();
}

export { TicTacToeClientScenario };

File: framework/languages/dotnet/samples/TicTacToe/Client/TicTacToeClientScenario.cs
using System.Runtime.CompilerServices;
using Microsoft.Extensions.Logging;
using Systems.Zlink.Stream.Connector.Contracts;
using TicTacToe.Shared.Contracts;
using Zlink.HttpClient;

namespace TicTacToe.Client;

public sealed class TicTacToeClientScenario(ILogger logger)
{
    // End-to-end client story:
    // 1. Create a game through HTTP and choose separate stream endpoints for host, guest, and observer.
    // 2. Connect the host and authenticate as player X.
    // 3. Connect the observer and subscribe to win milestone notifications.
    // 4. Join the host to the empty room, then connect the guest and verify join/start pushes.
    // 5. Play a deterministic move sequence where player X wins the top row.
    // 6. Verify the guest sees the final state and the observer receives the win milestone.
    public async ValueTask RunAsync(
        TicTacToeClientOptions options,
        CancellationToken cancellationToken = default)
    {
        using var api = ZLinkHttpClient.Create(options.ApiUrl.ToString())
            .Timeout(options.HttpTimeout)
            .Build();
        var room = api.Post("/games")
            .Body(new CreateGameHttpReq(options.GameName))
            .Async<CreateGameHttpRes>().AsTask().GetAwaiter().GetResult().Body;

        Ensure(!string.IsNullOrWhiteSpace(room.RoomId));
        Ensure(!string.IsNullOrWhiteSpace(room.OwnerPlayEndpoint));
        Ensure(room.PlayEndpoints.Count >= 2);
        Ensure(room.PlayEndpoints.Contains(room.OwnerPlayEndpoint, StringComparer.Ordinal));
        Ensure(room.PlayNodes.Count == room.PlayEndpoints.Count);
        Ensure(room.PlayNodes.Any(node =>
            string.Equals(node.StreamEndpoint, room.OwnerPlayEndpoint, StringComparison.Ordinal)));
        Ensure(room.RequiredLevel == 3);
        Ensure(room.GameName == options.GameName);

        var guestPlayEndpoint = room.PlayEndpoints.First(endpoint =>
            !string.Equals(endpoint, room.OwnerPlayEndpoint, StringComparison.Ordinal));
        var observerPlayEndpoint = guestPlayEndpoint;
        var observerPlayNode = room.PlayNodes.Single(node =>
            string.Equals(node.StreamEndpoint, observerPlayEndpoint, StringComparison.Ordinal));

        await using var client1 =
            TicTacToeClientConnections.CreateStreamClient(room.OwnerPlayEndpoint, options, "host", logger);
        await using var client2 =
            TicTacToeClientConnections.CreateStreamClient(guestPlayEndpoint, options, "guest", logger);
        await using var observer =
            TicTacToeClientConnections.CreateStreamClient(observerPlayEndpoint, options, "observer", logger);

        // Client 1 connects, authenticates as player X, and joins the empty room.
        await client1.Connect.Async(cancellationToken);

        var client1Authentication = await client1.Request(new AuthenticateReq(options.XActorId))
            .Async<AuthenticateRes>(cancellationToken);
        Ensure(client1Authentication.Player.ActorId == options.XActorId);
        Ensure(client1Authentication.Player.Level >= room.RequiredLevel);
        Ensure(client1Authentication.Player.Wins == 99);

        await observer.Connect.Async(cancellationToken);
        var observerAuthentication = await observer.Request(new AuthenticateReq(options.ObserverActorId))
            .Async<AuthenticateRes>(cancellationToken);
        Ensure(observerAuthentication.Player.ActorId == options.ObserverActorId);
        var observerSubscription =
            await observer.Request(new ObserveMilestoneReq()).Async<ObserveMilestoneRes>(cancellationToken);
        Ensure(observerSubscription.Subscribed);

        var client1Join = await client1.Request(new JoinGameReq(room.RoomId)).Async<JoinGameRes>(cancellationToken);
        Ensure(client1Join.State.RoomId == room.RoomId);
        Ensure(client1Join.State.Status == TicTacToeGameStatuses.WaitingForPlayers);
        Ensure(client1Join.State.XActorId == options.XActorId);
        Ensure(client1.ReceivedCount(nameof(PlayerJoinedNotify)) == 0);

        // Client 2 connects, authenticates as player O, and joins the same room.
        await client2.Connect.Async(cancellationToken);

        var client2Authentication = await client2.Request(new AuthenticateReq(options.OActorId))
            .Async<AuthenticateRes>(cancellationToken);
        Ensure(client2Authentication.Player.ActorId == options.OActorId);
        Ensure(client2Authentication.Player.Level >= room.RequiredLevel);
        Ensure(client2Authentication.Player.ActorId != client1Authentication.Player.ActorId);

        var client2Join = await client2.Request(new JoinGameReq(room.RoomId)).Async<JoinGameRes>(cancellationToken);
        Ensure(client2Join.State.RoomId == room.RoomId);
        Ensure(client2Join.State.Status == TicTacToeGameStatuses.InProgress);
        Ensure(client2Join.State.OActorId == options.OActorId);

        // Existing room members receive push packets when another player joins.
        var client1SawClient2Join = await client1.WaitFor<PlayerJoinedNotify>()
            .Where(message => message.Payload.ActorId == options.OActorId)
            .Async(cancellationToken);
        Ensure(client1SawClient2Join.Payload.ActorId == options.OActorId);
        Ensure(client1SawClient2Join.Payload.DisplayName == client2Authentication.Player.DisplayName);
        Ensure(client1SawClient2Join.Payload.Level == client2Authentication.Player.Level);
        Ensure(client1SawClient2Join.Payload.Mark == TicTacToeMarks.O);
        Ensure(client1SawClient2Join.Payload.RoomId == room.RoomId);
        Ensure(client1SawClient2Join.Payload.State.Status == TicTacToeGameStatuses.InProgress);
        Ensure(client2.ReceivedCount(nameof(PlayerJoinedNotify)) == 0);

        var client1SawGameStart = await client1.WaitFor<GameStateNotify>()
            .Where(message => message.Payload.State.Status == TicTacToeGameStatuses.InProgress
                              && message.Payload.State.OActorId == options.OActorId)
            .Async(cancellationToken);

        Ensure(client1SawGameStart.Payload.State.Status == TicTacToeGameStatuses.InProgress);
        Ensure(client1SawGameStart.Payload.State.OActorId == options.OActorId);
        Ensure(client1SawGameStart.Payload.State.NextTurn == TicTacToeMarks.X);

        // The move sequence is deterministic: client 1 completes the top row.
        var client1Move1 = await client1.Request(new PlaceMarkReq(0)).Async<PlaceMarkRes>(cancellationToken);
        Ensure(client1Move1.State.Board == "X........");
        Ensure(client1Move1.State.NextTurn == TicTacToeMarks.O);
        Ensure(client1Move1.State.LastMoveActorId == options.XActorId);
        Ensure(client1Move1.State.LastMoveCell == 0);

        var client2SawClient1Move1 = await client2.WaitFor<GameStateNotify>()
            .Where(message => message.Payload.State.LastMoveActorId == options.XActorId
                              && message.Payload.State.LastMoveCell == 0)
            .Async(cancellationToken);
        Ensure(client2SawClient1Move1.Payload.State.LastMoveActorId == options.XActorId);
        Ensure(client2SawClient1Move1.Payload.State.LastMoveCell == 0);
        Ensure(client2SawClient1Move1.Payload.State.Board == client1Move1.State.Board);

        var client2Move1 = await client2.Request(new PlaceMarkReq(3)).Async<PlaceMarkRes>(cancellationToken);
        Ensure(client2Move1.State.Board == "X..O.....");
        Ensure(client2Move1.State.NextTurn == TicTacToeMarks.X);
        Ensure(client2Move1.State.LastMoveActorId == options.OActorId);
        Ensure(client2Move1.State.LastMoveCell == 3);

        var client1SawClient2Move1 = await client1.WaitFor<GameStateNotify>()
            .Where(message => message.Payload.State.LastMoveActorId == options.OActorId
                              && message.Payload.State.LastMoveCell == 3)
            .Async(cancellationToken);
        Ensure(client1SawClient2Move1.Payload.State.LastMoveActorId == options.OActorId);
        Ensure(client1SawClient2Move1.Payload.State.LastMoveCell == 3);
        Ensure(client1SawClient2Move1.Payload.State.Board == client2Move1.State.Board);

        var client1Move2 = await client1.Request(new PlaceMarkReq(1)).Async<PlaceMarkRes>(cancellationToken);
        Ensure(client1Move2.State.Board == "XX.O.....");
        Ensure(client1Move2.State.NextTurn == TicTacToeMarks.O);
        Ensure(client1Move2.State.LastMoveActorId == options.XActorId);
        Ensure(client1Move2.State.LastMoveCell == 1);

        var client2SawClient1Move2 = await client2.WaitFor<GameStateNotify>()
            .Where(message => message.Payload.State.LastMoveActorId == options.XActorId
                              && message.Payload.State.LastMoveCell == 1)
            .Async(cancellationToken);
        Ensure(client2SawClient1Move2.Payload.State.LastMoveActorId == options.XActorId);
        Ensure(client2SawClient1Move2.Payload.State.LastMoveCell == 1);
        Ensure(client2SawClient1Move2.Payload.State.Board == client1Move2.State.Board);

        var client2Move2 = await client2.Request(new PlaceMarkReq(4)).Async<PlaceMarkRes>(cancellationToken);
        Ensure(client2Move2.State.Board == "XX.OO....");
        Ensure(client2Move2.State.NextTurn == TicTacToeMarks.X);
        Ensure(client2Move2.State.LastMoveActorId == options.OActorId);
        Ensure(client2Move2.State.LastMoveCell == 4);

        var client1SawClient2Move2 = await client1.WaitFor<GameStateNotify>()
            .Where(message => message.Payload.State.LastMoveActorId == options.OActorId
                              && message.Payload.State.LastMoveCell == 4)
            .Async(cancellationToken);
        Ensure(client1SawClient2Move2.Payload.State.LastMoveActorId == options.OActorId);
        Ensure(client1SawClient2Move2.Payload.State.LastMoveCell == 4);
        Ensure(client1SawClient2Move2.Payload.State.Board == client2Move2.State.Board);

        var client1FinalMove = await client1.Request(new PlaceMarkReq(2)).Async<PlaceMarkRes>(cancellationToken);
        Ensure(client1FinalMove.State.Board == "XXXOO....");
        Ensure(client1FinalMove.State.Status == TicTacToeGameStatuses.Won);
        Ensure(client1FinalMove.State.Winner == options.XActorId);
        Ensure(client1FinalMove.State.LastMoveActorId == options.XActorId);
        Ensure(client1FinalMove.State.LastMoveCell == 2);

        var client2SawFinal = await client2.WaitFor<GameStateNotify>()
            .Where(message => message.Payload.State.Status == TicTacToeGameStatuses.Won
                              && message.Payload.State.Winner == options.XActorId)
            .Async(cancellationToken);
        Ensure(client2SawFinal.Payload.State.Status == TicTacToeGameStatuses.Won);
        Ensure(client2SawFinal.Payload.State.Winner == options.XActorId);
        Ensure(client2SawFinal.Payload.State.Board == client1FinalMove.State.Board);

        var observerSawMilestone = await observer.WaitFor<WinMilestoneNotify>()
            .Where(message => message.Payload.ActorId == options.XActorId
                              && message.Payload.RoomId == room.RoomId)
            .Async(cancellationToken);
        Ensure(observerSawMilestone.Payload.DisplayName == client1Authentication.Player.DisplayName);
        Ensure(observerSawMilestone.Payload.Wins == 100);
        Ensure(observerSawMilestone.Payload.ReceivingSpotNodeRid == observerPlayNode.SpotNodeRid);
        logger.LogInformation(
            "observer-win-milestone=verified actor={0} wins={1} receivingSpotNodeRid={2}",
            observerSawMilestone.Payload.ActorId,
            observerSawMilestone.Payload.Wins,
            observerSawMilestone.Payload.ReceivingSpotNodeRid);

        client1.Send(new LeaveGameReq(room.RoomId)).Submit(cancellationToken);
        client2.Send(new LeaveGameReq(room.RoomId)).Submit(cancellationToken);
    }

    private static void Ensure(
        bool condition,
        [CallerArgumentExpression(nameof(condition))]
        string? expression = null)
    {
        if (!condition) throw new InvalidOperationException($"Ensure failed: {expression}");
    }
}
